### PR TITLE
[feature] Network Policy Resource

### PR DESCRIPTION
**Pull request by grahamplace**
Adds support for managing Snowflake Network Policy objects via the new `snowflake_network_policy` resource type.

## Test Plan

* [x] acceptance tests

* [x] tested built provider locally with trial Snowflake account

## References

* Issue: https://github.com/chanzuckerberg/terraform-provider-snowflake/issues/221
* https://docs.snowflake.com/en/sql-reference/sql/create-network-policy.html

### CODE REVIEW

File: pkg/provider/provider.go
Comment (ryanking) [inline pkg/provider/provider.go]: You probably need to run `make fmt` on your code.

File: pkg/resources/network_policy.go
Comment (ryanking) [inline pkg/resources/network_policy.go:78]: Comment needs to be escaped, see EscapeString func in the snowflake package.

File: pkg/resources/network_policy_acceptance_test.go
Comment (ryanking) [inline pkg/resources/network_policy_acceptance_test.go:52]: We should probably include testing adding to users here.

File: pkg/resources/network_policy_test.go
Comment (ryanking) [inline pkg/resources/network_policy_test.go:37]: Can we fix this before merging?

File: pkg/resources/network_policy.go
Comment (ryanking) [inline pkg/resources/network_policy.go:85]: Can we move this sql generation to the snowflake package with the rest of the SQL?

**Review by ryanking — CHANGES_REQUESTED** [review]
Thanks! This looks great.

A few minor bits of feedback and this should be good to go.

Thanks for writing acceptance tests.

**Comment by ryanking** [comment]
Note that I am seeing some acceptance test failures, running `make test-acceptance` locally -

```
=== RUN   TestAccNetworkPolicy
    testing.go:683: Step 0 error: errors during apply:
        
        Error: 000606 (57P03): No active warehouse selected in the current session.  Select an active warehouse with the 'use warehouse' command.
        
        
          on /var/folders/ly/qwpw0g113mz15651mkc9wd400000gr/T/tf-test836548438/main.tf line 2:
          (source code not available)
        
        
    testing.go:744: Error destroying resource! WARNING: Dangling resources
        may exist. The full state and error is shown below.
        
        Error: errors during refresh: 000606 (57P03): No active warehouse selected in the current session.  Select an active warehouse with the 'use warehouse' command.
        
        
        State: <nil>
```

It seems that the select query needs a warehouse, which means we would need to add that to the provider configuration. In the past we've mostly avoided that and instead dealt with the the annoyances of working with just `show` and `describe`, which don't need a warehouse.

File: pkg/resources/network_policy_acceptance_test.go
Comment (grahamplace) [reply pkg/resources/network_policy_acceptance_test.go:58]: I've added a `users` list to the test network policy resource definition (and a dummy user resource definition) so at least the user-granting code should be executing during this acceptance test now.

However, (as far as I can tell) I have to define:
`ImportStateVerifyIgnore: []string{"set_for_account", "users"},` 

because nowhere in `ReadNetworkPolicy` is there:
`data.Set("set_for_account", ...)` or
`data.Set("users", ...)`

This is the case because there doesn't appear to be a way in Snowflake to do "get all users with this Network Policy set" or "tell me if this network policy is set for the Snowflake account". Neither `SHOW` nor `DESC` returns any information about the active/applied status of a given Network Policy:

- Show: https://docs.snowflake.com/en/sql-reference/sql/show-network-policies.html
- Describe: https://docs.snowflake.com/en/sql-reference/sql/desc-network-policy.html

Is it acceptable to leave those out of `ReadNetworkPolicy` (and thus having to include them in `ImportStateVerifyIgnore`)?

Comment (grahamplace) [reply pkg/resources/network_policy_acceptance_test.go:58]: FWIW this is why I had to write this resource with `ForceNew: true` for those two fields, with the "exclusive assignment" explanation in the docs — the simplest way I could figure out how to deal with this is to just always destroy/recreate and re-grant to users or re-set on account

**Comment by grahamplace** [comment]
> Note that I am seeing some acceptance test failures, running `make test-acceptance` locally -
> 
> ```
> === RUN   TestAccNetworkPolicy
>     testing.go:683: Step 0 error: errors during apply:
>         
>         Error: 000606 (57P03): No active warehouse selected in the current session.  Select an active warehouse with the 'use warehouse' command.
>         
>         
>           on /var/folders/ly/qwpw0g113mz15651mkc9wd400000gr/T/tf-test836548438/main.tf line 2:
>           (source code not available)
>         
>         
>     testing.go:744: Error destroying resource! WARNING: Dangling resources
>         may exist. The full state and error is shown below.
>         
>         Error: errors during refresh: 000606 (57P03): No active warehouse selected in the current session.  Select an active warehouse with the 'use warehouse' command.
>         
>         
>         State: <nil>
> ```
> 
> It seems that the select query needs a warehouse, which means we would need to add that to the provider configuration. In the past we've mostly avoided that and instead dealt with the the annoyances of working with just `show` and `describe`, which don't need a warehouse.

Similar to my reply on your other comment — the issue here is that Snowflake's implementation of Network Policies is weird/somewhat different than other object types. 

More specifically in this case, I think we _have_ to do the `RESULT_SCAN` approach I used (which is the cause of acceptance tests user requiring a warehouse). This is because there is no way to `SHOW`/`DESC` the information about a _single_ Network Policy in Snowflake SQL:

- [DESC](https://docs.snowflake.com/en/sql-reference/sql/desc-network-policy.html) is limited to a single Network Policy, but it only returns the IP address lists: 
image

- [SHOW](https://docs.snowflake.com/en/sql-reference/sql/show-network-policies.html) _only_ allows `SHOW NETWORK POLICIES` — for some reason, unlike other object types, the output cannot be limited to a single Network Policy object via `LIKE`. Adding a `LIKE` does not limit the returned data:
image

Thus, in order to get the current `comment` on the Network Policy in question, the `SHOW` output has to be cut down to just `name = "mypolicy"` via a second query that uses `RESULT_SCAN('original_query_id')`


I worked around this by setting `DEFAULT_WAREHOUSE` on the Travis user. Not ideal obviously, but I don't have other ideas for how to implement `ReadNetworkPolicy` without the `RESULT_SCAN` approach given limitations outlined above. Open to ideas!

**Comment by grahamplace** [comment]
^ while writing that I realized that I can probably just refactor to do the SHOW/DESC post-processing in Go instead of in Snowflake SQL, thus bypassing the `RESULT_SCAN` and the need for a Warehouse

Still posted as it's useful context, but I'll take a stab at that refactor now

File: pkg/resources/network_policy_acceptance_test.go
Comment (ryanking) [reply pkg/resources/network_policy_acceptance_test.go:87]: Ah that makes sense!

Thinking about this a bit more, I think we should consider separating the attachment to a separate resource. This should allow us to avoid the scenario where we need to recreate the network policy resource every time one of the attachments changes. Of course the attachments may need to be recreated, but that might be tolerable. What do you think?

Comment (grahamplace) [reply pkg/resources/network_policy_acceptance_test.go:87]: I'm on board with this — a separate attachment resource was my original plan, but I moved away from it because I (erroneously) observed that if you dropped a network policy, the users with that policy set continued to be restricted by the policy. Hence, I thought we needed to keep the attachment tightly coupled with the resource. 

However, I tested this understanding again and it turns out there's just a bit of lag (not more than a minute) between dropping the network policy and the user then being able to open a session from any IP address

So this approach _should_ work. I'll update PR

Comment (grahamplace) [reply pkg/resources/network_policy_acceptance_test.go:87]: I guess one thing to think about is that if we do this, the `snowflake_network_policy_attachment` resource wouldn't behave like the similar *_grant resources (e.g. warehouse/database/etc) and create exclusive attachments

Because there's not a way of getting existing attachments from `SHOW`/`DESC`ing network policies/users/accounts, a network policy could be attached to a user outside of Terraform, and then changes to the `snowflake_network_policy_attachment` wouldn't destroy that attachment. That's probably fine? We should just document it

Comment (ryanking) [reply pkg/resources/network_policy_acceptance_test.go:87]: Yeah I think that is fine given the limitations we are working with.

**Comment by aidan-melen** [comment]
thanks for working on this feature @grahamplace and @ryanking!

**Review by alldoami — APPROVED** [review]
Ran `test-acceptancec-ci` locally and passed all the test! Thanks for the work! 🥳  I'll let @ryanking take one more look, but LGTM!

**Comment by grahamplace** [comment]
Thanks for the review @alldoami! Per Ryan's [suggestion](https://github.com/chanzuckerberg/terraform-provider-snowflake/pull/271#discussion_r504157297), I still need to refactor this to move the attachment into a separate resource. That work is on my todo list for today, so hopefully I can put that change up and tag you both for another look

File: pkg/resources/network_policy_acceptance_test.go
Comment (grahamplace) [reply pkg/resources/network_policy_acceptance_test.go]: Done — the one thing I'm unsure about is `ReadNetworkPolicyAttachment` — given above limitations, there's no way to actually implement `schema.ReadFunc` for the attachment resource. I don't think I actually need `ReadNetworkPolicyAttachment` implemented for this, but `InternalValidate` fails without it. To get around that, I just implemented a dummy func 🤷 

lmk if there's a better way to handle this!

**Comment by grahamplace** [comment]
Done! Ready for another look @ryanking @alldoami 🙂  Would appreciate input on this bit:
https://github.com/chanzuckerberg/terraform-provider-snowflake/pull/271#discussion_r508133483

File: pkg/resources/network_policy_acceptance_test.go
Comment (ryanking) [inline pkg/resources/network_policy_acceptance_test.go:34]: Is there a reason for this to be false?

**Review by ryanking — CHANGES_REQUESTED** [review]
I am currently getting acceptance test failures too:

```
pkg/resources/network_policy_attachment_acceptance_test.go:12:6: TestAccNetworkPolicyAttachment redeclared in this block
	previous declaration at pkg/resources/network_policy_acceptance_test.go:12:40
pkg/resources/network_policy_attachment_acceptance_test.go:40:6: networkPolicyAttachmentConfig redeclared in this block
	previous declaration at pkg/resources/network_policy_acceptance_test.go:40:55
```

**Comment by grahamplace** [comment]
> I am currently getting acceptance test failures too:
> 
> ```
> pkg/resources/network_policy_attachment_acceptance_test.go:12:6: TestAccNetworkPolicyAttachment redeclared in this block
> 	previous declaration at pkg/resources/network_policy_acceptance_test.go:12:40
> pkg/resources/network_policy_attachment_acceptance_test.go:40:6: networkPolicyAttachmentConfig redeclared in this block
> 	previous declaration at pkg/resources/network_policy_acceptance_test.go:40:55
> ```

Ah oops, I accidentally overwrote `network_policy_acceptance_test.go` with `network_policy_attachment_acceptance_test.go`, should be corrected now

File: pkg/resources/network_policy_acceptance_test.go
Comment (grahamplace) [reply pkg/resources/network_policy_acceptance_test.go]: I set this to false while developing because I wasn't able to get it to work, I thought due to the inability to _actually_ implement `ReadNetworkPolicyAttachment`

But it looks like it works now so switching it back to true

Comment (grahamplace) [reply pkg/resources/network_policy_acceptance_test.go]: annnd I was wrong -- it's actually still not working:
```
=== RUN   TestAccNetworkPolicyAttachment
    TestAccNetworkPolicyAttachment: testing.go:683: Step 1 error: ImportStateVerify attributes not equivalent. Difference is shown below. Top is actual, bottom is expected.

        (map[string]string) {
        }


        (map[string]string) (len=5) {
         (string) (len=19) "network_policy_name": (string) (len=10) "uowauijhoy",
         (string) (len=15) "set_for_account": (string) (len=5) "false",
         (string) (len=7) "users.#": (string) (len=1) "2",
         (string) (len=16) "users.2494182523": (string) (len=10) "test-user1",
         (string) (len=16) "users.3213322168": (string) (len=10) "test-user2"
        }

--- FAIL: TestAccNetworkPolicyAttachment (7.03s)
```

I think because `ReadNetworkPolicyAttachment` isn't really implemented?

**Review by ryanking — APPROVED** [review]

File: pkg/resources/network_policy_acceptance_test.go
Comment (grahamplace) [reply pkg/resources/network_policy_acceptance_test.go]: Just confirming you saw this last comment/[commit](https://github.com/chanzuckerberg/terraform-provider-snowflake/pull/271/commits/53a3f9a1e452f6f78245cb9de1282aec69ef890b) @ryanking? Does the merge mean you're fine skipping `ImportStateVerify` given Snowflake limitations that prevent us from show/desc'ing anything about a policy attachment on users/account?

Comment (ryanking) [reply pkg/resources/network_policy_acceptance_test.go]: Yeah its fine.

Comment (grahamplace) [reply pkg/resources/network_policy_acceptance_test.go]: Awesome, thanks for all the help getting this merged!

**Pull request merged**
